src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel.cs

```diff
@@ -0,0 +1,289 @@
+                    DocumentSymbolViewModelItems = documentSymbolViewModelItems;
+                }
+
+                if (searchText is { } currentQuery)
+                {
+                    ImmutableArray<DocumentSymbolDataViewModel> documentSymbolViewModelItems;
+                    if (currentQuery == string.Empty)
+                    {
+                        // search was cleared, show all data.
+                        documentSymbolViewModelItems = DocumentOutlineHelper.GetDocumentSymbolItemViewModels(model.DocumentSymbolData);
+                        ApplyExpansionStateToNewItems(documentSymbolViewModelItems, DocumentSymbolViewModelItems);
+                    }
+                    else
+                    {
+                        // We are going to show results so we unset any expand / collapse state
+                        documentSymbolViewModelItems = DocumentOutlineHelper.GetDocumentSymbolItemViewModels(
+                             DocumentOutlineHelper.SearchDocumentSymbolData(model.DocumentSymbolData, currentQuery, cancellationToken));
+                    }
+
+                    DocumentSymbolViewModelItems = documentSymbolViewModelItems;
+                }
+
+                if (position is { } currentPosition)
+                {
+                    var caretPoint = currentPosition.Point.GetPoint(_textBuffer, PositionAffinity.Predecessor);
+                    if (!caretPoint.HasValue)
+                    {
+                        // document has changed since we were queued to the point that we can't find this position.
+                        return;
+                    }
+
+                    var symbolToSelect = DocumentOutlineHelper.GetDocumentNodeToSelect(DocumentSymbolViewModelItems, model.OriginalSnapshot, caretPoint.Value);
+                    if (symbolToSelect is null)
+                    {
+                        return;
+                    }
+
+                    DocumentOutlineHelper.ExpandAncestors(DocumentSymbolViewModelItems, symbolToSelect.SelectionRangeSpan);
+                    symbolToSelect.IsSelected = true;
+                }
+            }
+
+            // If we aren't filtering to search results do expand/collapse
+            if (expansion is { } expansionOption && string.IsNullOrEmpty(searchText))
+            {
+                DocumentOutlineHelper.SetExpansionOption(DocumentSymbolViewModelItems, expansionOption);
+            }
+            else if (expansion is ExpansionOption.Collapse)
+            {
+                DocumentOutlineHelper.SetExpansionOption(DocumentSymbolViewModelItems, ExpansionOption.Collapse);
+            }
+
+            static void ApplyExpansionStateToNewItems(ImmutableArray<DocumentSymbolDataViewModel> oldItems, ImmutableArray<DocumentSymbolDataViewModel> newItems)
+            {
+                if (DocumentOutlineHelper.AreAllTopLevelItemsCollapsed(oldItems))
```

Comment:
does anything else call AreAllTopLevelItemsCollapsed?  if not, move here.  

note: the Empty check inside of AreAllTopLevelItemsCollapsed seems unnecessary. if yo uhave no items, then calling into either of these branches seems fine.